sql/src/main/java/io/crate/operation/projectors/ShardingShardRequestConsumer.java

```diff
@@ -129,41 +131,73 @@ public ShardingShardRequestAccumulator(ClusterService clusterService,
     }
 
     @Override
-    public void onItem(Row row) {
+    public CompletableFuture<? extends Iterable<Row>> apply(BatchIterator batchIterator) {
+        CompletableFuture<List<Row>> result = new CompletableFuture<>();
+        consumeIterator(batchIterator, result);
+        return result;
+    }
+
+    private void consumeIterator(BatchIterator batchIterator, CompletableFuture<List<Row>> result) {
+        Row row = RowBridging.toRow(batchIterator.rowData());
+        try {
+            while (batchIterator.moveNext()) {
+                onRow(row);
+                if (idxWithinBatch == batchSize) {
+                    CompletableFuture<BitSet> executeBatchFuture = execute(false);
+                    executeBatchFuture.whenComplete((r, t) -> {
+                        if (t == null) {
+                            consumeIterator(batchIterator, result);
+                        } else {
+                            result.completeExceptionally(t);
+                        }
+                    });
+                    idxWithinBatch = 0;
+                }
+            }
+
+            if (batchIterator.allLoaded()) {
+                batchIterator.close();
+                execute(true).whenComplete((r, t) -> {
+                    if (t == null) {
+                        result.complete(Collections.singletonList(new Row1(Long.valueOf(r.cardinality()))));
```

Comment:
I mean `new Row1((long)r.cardinality())` - the Long.valueOf variant is afaik a warning in intellij with default settings